homeassistant/components/tado/device_tracker.py

```diff
@@ -1,149 +1,132 @@
 """Support for Tado Smart device trackers."""
 from __future__ import annotations
 
-import asyncio
-from collections import namedtuple
-from datetime import timedelta
-from http import HTTPStatus
 import logging
+from typing import Any
 
-import aiohttp
-import voluptuous as vol
-
-from homeassistant.components.device_tracker import (
-    DOMAIN,
-    PLATFORM_SCHEMA as BASE_PLATFORM_SCHEMA,
-    DeviceScanner,
-)
-from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
-from homeassistant.core import HomeAssistant
-from homeassistant.helpers.aiohttp_client import async_create_clientsession
-import homeassistant.helpers.config_validation as cv
-from homeassistant.helpers.typing import ConfigType
-from homeassistant.util import Throttle
-
-_LOGGER = logging.getLogger(__name__)
-
-CONF_HOME_ID = "home_id"
-
-MIN_TIME_BETWEEN_SCANS = timedelta(seconds=30)
-
-PLATFORM_SCHEMA = BASE_PLATFORM_SCHEMA.extend(
-    {
-        vol.Required(CONF_USERNAME): cv.string,
-        vol.Required(CONF_PASSWORD): cv.string,
-        vol.Optional(CONF_HOME_ID): cv.string,
-    }
-)
-
-
-def get_scanner(hass: HomeAssistant, config: ConfigType) -> TadoDeviceScanner | None:
-    """Return a Tado scanner."""
-    scanner = TadoDeviceScanner(hass, config[DOMAIN])
-    return scanner if scanner.success_init else None
-
-
-Device = namedtuple("Device", ["mac", "name"])
+from homeassistant.components.device_tracker import SourceType
+from homeassistant.components.device_tracker.config_entry import ScannerEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
 
+from .const import DATA, DOMAIN, SIGNAL_TADO_MOBILE_DEVICE_UPDATE_RECEIVED
 
-class TadoDeviceScanner(DeviceScanner):
-    """Scanner for geofenced devices from Tado."""
+_LOGGER = logging.getLogger(__name__)
 
-    def __init__(self, hass, config):
-        """Initialize the scanner."""
-        self.hass = hass
-        self.last_results = []
 
-        self.username = config[CONF_USERNAME]
-        self.password = config[CONF_PASSWORD]
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Tado device scannery entity."""
+    _LOGGER.debug("Setting up Tado device scanner entity")
+    tado = hass.data[DOMAIN][entry.entry_id][DATA]
+    tracked: set = set()
+
+    @callback
+    async def async_update_devices() -> None:
```

Comment:
Please go over all the functions and don't mix `@callback` with `async def`.

The functions that don't need to `await` inside should use `@callback`. If `await` is needed, use `async def`.

Reply:
Update provided. :)

Reply:
This is still a coroutine function while there is no await inside. Please make it a callback.